aten/src/ATen/native/quantized/cpu/conv_serialization.h

```diff
@@ -0,0 +1,292 @@
+// the current format.
+template <uint32_t kSpatialDim>
+ConvParamsSerializationType parse_conv_serialized_state(c10::IValue v) {
+
+  // determine the version based on IValue contents
+  int version = -1;
+  if (v.isTuple()) {
+    auto elements = v.toTuple()->elements();
+    if (elements.size() > 0) {
+      auto firstElement = elements[0];
+      if (firstElement.isTensor()) {
+        version = 1;
+      } else if (firstElement.isString()) {
+        std::string version_str = firstElement.toStringRef();
+        // note: not parsing the string to automatically handle bad
+        // inputs
+        if (version_str == "2") {
+          version = 2;
+        }
+      }
+    }
+  }
+  TORCH_INTERNAL_ASSERT(version != -1, "Unable to parse serialization version");
+
+  ConvParamsSerializationType state;
+
+  if (version == 1) {
+    // version 1 - convert to version 2 manually
+
+    auto elements = v.toTuple()->elements();
+
+    at::Tensor weight = elements[0].toTensor();
+    c10::optional<at::Tensor> bias = elements[1].toOptional<at::Tensor>();
+    torch::List<at::Tensor> stride_x_kSpatialDim = elements[2].toTensorList();
+    torch::List<at::Tensor> padding_x_kSpatialDim = elements[3].toTensorList();
+    torch::List<at::Tensor> dilation_x_kSpatialDim = elements[4].toTensorList();
+    at::Tensor groups = elements[5].toTensor();
+
+    std::string version = "2";
+    std::vector<at::Tensor> non_optional;
+    std::vector<c10::optional<at::Tensor>> optional;
+
+    std::vector<int8_t> params_vec;
+    params_vec.push_back(kSpatialDim);
+    for (int i = 0; i < stride_x_kSpatialDim.size(); i++) {
+      auto stride = stride_x_kSpatialDim.get(i);
+      params_vec.push_back(stride[0].item<int8_t>());
+    }
+    for (int i = 0; i < padding_x_kSpatialDim.size(); i++) {
+      auto padding = padding_x_kSpatialDim.get(i);
+      params_vec.push_back(padding[0].item<int8_t>());
+    }
+    for (int i = 0; i < dilation_x_kSpatialDim.size(); i++) {
+      auto dilation = dilation_x_kSpatialDim.get(i);
+      params_vec.push_back(dilation[0].item<int8_t>());
+    }
+    // output_padding does not exist in v1, so we fill in a default value
+    for (int i = 0; i < kSpatialDim; i++) {
+      params_vec.push_back(0);
+    }
+    params_vec.push_back(groups[0].item<int8_t>());
+    // transpose does not exist in v1, so we fill in a default value
+    params_vec.push_back(0);
+    int64_t vec_size = params_vec.size();
+    at::Tensor params_tensor = at::from_blob(params_vec.data(),
+        {vec_size}, at::TensorOptions().dtype(at::kChar))
+      // clone to retain ownership of the data
+      .clone();
+
+    non_optional.emplace_back(std::move(params_tensor));
+    non_optional.emplace_back(std::move(weight));
+    optional.emplace_back(std::move(bias));
+
+    state = std::tie(version, non_optional, optional);
+
+  } else {
+    // version 2
+    state = v.to<ConvParamsSerializationType>();
+  }
+
+  return state;
```

Comment:
I would suggest writing this as follows to reduce the risk of accidentally returning an empty result if this code changes in the future:

```

    return std::tie(version, non_optional, optional);

  } else if (version == 2) {
    // version 2
    return v.to<ConvParamsSerializationType>();

  } else {
    TORCH_INTERNAL_ASSERT(false, "Unexpected serialized qconv version: ", version);
  }
}
```